docs/design/fabric-v2+/fpc-lifecycle-v2-new.puml

```diff
@@ -0,0 +1,139 @@
+/'
+    Copyright 2020 Intel Corporation
+    Copyright IBM Corp. All Rights Reserved.
+
+    SPDX-License-Identifier: Apache-2.0
+'/
+
+@startuml
+
+!pragma teoz true
+
+hide footbox
+title Foot Box removed
+
+title Fabric Private Chaincode - Lifecycle v2
+
+box "Org1"
+actor	Admin order 10
+entity "FPC_Peer1_Wrapper" as FPC_Peer1_Wrapper order 15
+participant Peer1 order 20
+entity "FPC_Peer2_Wrapper" as FPC_Peer2_Wrapper order 30
+participant Peer2 order 40
+end box
+
+box "Channel Abstraction"
+participant Channel order 70
+note over Channel
+    This box abstracts all other orgs and components.
+    Actions are supposed to be performed by channel participants.
+end note
+end box
+
+entity Orderer order 80
+
+note over Admin
+    Admin uses the fpc-peer wrappers, which abstract the Fabric Peer CLI.
+end note
+
+group chaincode package
+Admin -> FPC_Peer1_Wrapper : lifecycle chaincode package
+note right
+    **Chaincode package**
+    as tar.gz with
+    metadata.json = <path:(enclaves_path), type:(cc_lang), label:(cc_label), sgx_mode: (sim/hw)>
+    code.tar.gz : compressed file with enclave binary and mrenclave
+end note
+return
+end /' group chaincode package '/
+
+group install
+Admin -> FPC_Peer1_Wrapper ++: lifecycle chaincode install
+FPC_Peer1_Wrapper -> Peer1 ++: lifecycle chaincode install
+return packageId
+note over FPC_Peer1_Wrapper
+    Bookeeping operations to take into account that:
+    - FPC's initial version will run a chaincode on a designated peer
+    - the install lifecyle op is peer-specific
+    - the approveformyorg and commit ops are org-specific
+    So the approve stores data for an FPC chaincode installed locally,
+    and the commit checks for data of a locally-approved FPC chaincode.    
+end note
+FPC_Peer1_Wrapper -> FPC_Peer1_Wrapper : retrieve cc_language from package
+note over FPC_Peer1_Wrapper
+    Queryinstalled is redudant, yet consistent with implementation.    
+end note
+FPC_Peer1_Wrapper -> Peer1 ++: lifecycle chaincode queryinstalled
+return packageId
+FPC_Peer1_Wrapper -> FPC_Peer1_Wrapper : if cc_language==fpc then store packageId
+return packageId /' Admin -> FPC_Peer1_Wrapper '/
+end /' group install '/
+
+ref over Admin, FPC_Peer2_Wrapper, Peer2
+    install chaincode on Peer2
+end /' ref '/
+
+group /' approve '/
+Admin -> FPC_Peer1_Wrapper ++: lifecycle chaincode approveformyorg
+FPC_Peer1_Wrapper -> Peer1 ++: lifecycle chaincode approveformyorg
+note right
+    **Chaincode definition**
+    Name: cc_name, Version: mrenclave,
```

Comment:
btw: currently the peer cli takes the version passed by the humans as-is and doesn't do anything with mrenclave.  Given that the package defines the MRenclave, i think we probably should just skip the version parameter from `approveformyorg` and automatically set it from the package or at least cross-check?  (if we keep track of sequence numbers in the bookkeeping, we could also drop it from `commit` ...)

Reply:
Depends a bit on the workflow and ground-truth you expect the users to use.  I was assuming the code would be the reference for the user as easier but it is true that even though more cumbersome for the user (as well as for the scripting!) the ultimate ground truth is MRENCLAVE.

Given that wrong versions are quite likely (e.g., typo, copy/paste errros ...) the error happens only delayed (and probably also not very intuitively), cross-checking at approve time would seem a good idea.
  
BTW: I think it wouldn't be TLCC but rather ERCC would find the problem? (as discussed elsewhere ERCC would handle attestation and TLCC defers to it by just verifying proper ERCC transactions have happened. In that case I would expect TLCC only consult ercc state to see whether the enclave endorsement is done by a properly registered key?)